spec/services/bulk_submission_service_spec.rb

```diff
@@ -0,0 +1,238 @@
+            nino: "JA654321D",
+          ),
+        ])
+      end
+    end
+
+    context "when bulk submission file has whitespace in csv content" do
+      let(:bulk_submission) do
+        create(:bulk_submission,
+               :with_content_for_original_file,
+               content_for_original_file: content)
+      end
+
+      let(:content) do
+        <<~CSV
+          period_start_date    , period_end_date,first_name \t, last_name  \t, \tdate_of_birth    , nino
+          01/01/2023      , 01/03/2023,  Jim   ,  Bob ,   01/01/2001 ,  JA123456D
+        CSV
+      end
+
+      it "creates submissions with expected attributes" do
+        expect { call }.to change(Submission, :count).by(2)
+
+        submissions = Submission.all.map {|e| e.attributes.with_indifferent_access }
+
+        expect(submissions).to match_array([
+          hash_including(
+            use_case: "one",
+            period_start_at: Date.parse("01/01/2023"),
+            period_end_at: Date.parse("01/03/2023"),
+            first_name: "Jim",
+            last_name: "Bob",
+            dob: Date.parse("01/01/2001"),
+            nino: "JA123456D",
+          ),
+          hash_including(
+            use_case: "two",
+            period_start_at: Date.parse("01/01/2023"),
+            period_end_at: Date.parse("01/03/2023"),
+            first_name: "Jim",
+            last_name: "Bob",
+            dob: Date.parse("01/01/2001"),
+            nino: "JA123456D",
+          ),
+        ])
+      end
+    end
+
+    context "when bulk submission has no file attached" do
+      let(:bulk_submission) { create(:bulk_submission) }
+
+       it { expect { call }.to raise_error ArgumentError, /Cannot parse nil/ }
+    end
+
+    context "when bulk submission file has record with invalid date format" do
+      let(:bulk_submission) do
+        create(:bulk_submission,
+               :with_content_for_original_file,
+               content_for_original_file: content)
+      end
+
+      before do
+        allow(Rails.logger).to receive(:error)
+      end
+
+      context "with invalid start_date" do
+        let(:content) do
+          <<~CSV
+            period_start_date, period_end_date, first_name, last_name, date_of_birth, nino
+            AA, 01/03/2023, Jim, Bob, 01/01/2001, JA123456D
+          CSV
+        end
+
+        it "logs but does not raise error" do
+          expect { call }.not_to raise_error
+          expect(Rails.logger)
+            .to have_received(:error)
+            .with("invalid date for period_start_at")
+            .twice
+        end
+      end
+
+      context "with invalid end_date" do
+        let(:content) do
+          <<~CSV
+            period_start_date, period_end_date, first_name, last_name, date_of_birth, nino
+            01/01/2023, AA, Jim, Bob, 01/01/2001, JA123456D
+          CSV
+        end
+
+        it "logs error, does not raise, does not create submission for use_case one and two" do
```

Comment:
Are we actually testing that it hasn't created a submission here? Should we be?

Reply:
we should for completeness IMO